scio-core/src/main/scala/com/spotify/scio/testing/TestDataManager.scala

```diff
@@ -119,12 +127,70 @@ private[scio] class TestDistCache(val m: Map[DistCacheIO[_], _]) {
   }
 }
 
+private[scio] class TestTransform(val m: Map[String, Map[_, _]]) {
```

Comment:
Hm... instead of this `TestTransform` class, I think it would be more idiomatic+flexible to work with actual `PTransform`s as much as possible. What if we had a utility class somewhere in the `testing` package that created mock transforms? Like:

```scala
object TransformMocks {
  def forAsyncLookupDoFn[InputT, OutpuT](m: Map[InputT, OutputT]): PTransform[PCollection[InputT], PCollection[KV[InputT, Try[OutputT]]] = {
     // just use a Beam `MapElements` transform to implement
  }

  // mock for Source transforms, could be useful for, say, SMB
  def forSource[InputT](s: Seq[InputT]): PTransform[PBegin, PCollection[InputT]] = {
     // use Beam `Create` transform
  }

  // etc...
}
```

then the `mockTransform` method you added to JobTest could take in an actual `PTransform` as input and use the `Pipeline#replaceAll` method to drop it in, as Julien suggested. (Would prob be clearer to rename from `mockTransform` to `replaceTransform` in that case). wdyt?